src/node_nanomsg.cc

```diff
@@ -191,35 +190,40 @@ NAN_METHOD(Err) {
   info.GetReturnValue().Set(Nan::New(nn_strerror(nn_errno())).ToLocalChecked());
 }
 
-typedef struct nanomsg_socket_s {
-  uv_poll_t poll_handle;
-  uv_os_sock_t sockfd;
-  Nan::Callback *callback;
-} nanomsg_socket_t;
+class NanoMsgPollCtx {
+  private:
+  public:
+    uv_poll_t poll_handle;
+    uv_os_sock_t sockfd;
+    Nan::Callback *callback;
+    NanoMsgPollCtx (int s, bool is_sender, Local<v8::Function> cb) {
+      callback = new Nan::Callback(cb);
+      size_t siz = sizeof(uv_os_sock_t);
+      nn_getsockopt(s, NN_SOL_SOCKET, is_sender ? NN_SNDFD : NN_RCVFD, &sockfd,
+          &siz);
+      poll_handle.data = this;
+    }
+    ~NanoMsgPollCtx () {
+      delete callback;
+    }
+};
 
-void NanomsgReadable(uv_poll_t *req, int status, int events) {
+static void NanomsgReadable(uv_poll_t *req, int status, int events) {
   Nan::HandleScope scope;
 
-  nanomsg_socket_t *context;
-  context = reinterpret_cast<nanomsg_socket_t *>(req);
+  NanoMsgPollCtx *context = reinterpret_cast<NanoMsgPollCtx*>(req->data);
 
   if (events & UV_READABLE) {
     Local<Value> argv[] = { Nan::New<Number>(events) };
     context->callback->Call(1, argv);
   }
 }
 
-NAN_METHOD(PollSendSocket) {
+NAN_METHOD(PollSocket) {
   int s = Nan::To<int>(info[0]).FromJust();
-  Nan::Callback *callback = new Nan::Callback(info[1].As<Function>());
-
-  nanomsg_socket_t *context;
-  size_t siz = sizeof(uv_os_sock_t);
-
-  context = reinterpret_cast<nanomsg_socket_t *>(calloc(1, sizeof *context));
-  context->poll_handle.data = context;
-  context->callback = callback;
-  nn_getsockopt(s, NN_SOL_SOCKET, NN_SNDFD, &context->sockfd, &siz);
+  bool is_sender = Nan::To<bool>(info[1]).FromJust();
```

Comment:
I assume there is enough sanity checking on the JavaScript side to ensure this is not `Nothing`.